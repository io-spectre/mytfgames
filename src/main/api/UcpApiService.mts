import { makeServiceIdentifier } from "$main/utils";

export enum LoginState {
  Unknown = -1,
  NotLoggedIn,
  LoggedIn,
}

interface UcpApiService {
  checkLoginState(cancel?: AbortSignal | undefined): Promise<LoginState>;
}
const UcpApiService = makeServiceIdentifier<UcpApiService>("ucp api service");
export { UcpApiService };
