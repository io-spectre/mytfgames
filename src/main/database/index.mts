export * from "./AppDatabase.mjs";
export * from "./GameTable.mjs";
