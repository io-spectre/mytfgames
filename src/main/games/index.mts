export * from "./GameDataService.mjs";
export * from "./GameVersionService.mjs";
