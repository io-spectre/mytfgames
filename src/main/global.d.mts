export {}; // Make this a module

declare global {
  /** @deprecated this global variable is for debugging purposes only */
  var main: unknown;
}
