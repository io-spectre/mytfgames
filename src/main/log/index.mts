export * from "./annotations.mjs";
export * from "./constants.mjs";
