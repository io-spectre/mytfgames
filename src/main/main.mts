/// <reference types="vite/client" />
import "reflect-metadata/lite";

import { app } from "electron";
import { Container } from "inversify";

import { MainApp } from "$main/app";
import { makeServiceIdentifier } from "$main/utils";

import { AppModule } from "./app/app.module.mjs";
import { PalBrowserModule } from "./pal/browser.module.mjs";

const container = new Container();
container.load(
  // keep this list sorted
  AppModule,
  PalBrowserModule,
);

const mainAppTypeId = makeServiceIdentifier<MainApp>("main app");
container.bind(mainAppTypeId).to(MainApp).inSingletonScope();
const mainApp = (globalThis.main = container.get(mainAppTypeId));

mainApp.run().catch((error: unknown) => {
  console.error(`MainApp.run() failed`, error);
  app.exit(-1);
});
