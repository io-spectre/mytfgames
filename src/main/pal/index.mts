export * from "./browser/index.mjs";
export * from "./dialogs/index.mjs";
export * from "./Ipc/index.mjs";
export * from "./JobSchedule.mjs";
