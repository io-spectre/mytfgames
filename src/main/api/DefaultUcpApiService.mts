import { load as cheerioLoad } from "cheerio";
import type { net } from "electron/main";
import { injectable } from "inversify";

import { LoginState, UcpApiService } from "./UcpApiService.mjs";

@injectable()
export class DefaultUcpApiService implements UcpApiService {
  #net: Pick<typeof net, "fetch">;
  constructor(_net: Pick<typeof net, "fetch">) {
    this.#net = _net;
  }

  async checkLoginState(cancel?: AbortSignal): Promise<LoginState> {
    cancel?.throwIfAborted();
    try {
      const abortController = new AbortController();
      if (cancel) {
        const onCancel = function onCancel(this: AbortSignal) {
          abortController.abort(this.reason);
        };
        cancel.addEventListener("abort", onCancel, { once: true });
        abortController.signal.addEventListener(
          "abort",
          () => {
            cancel.removeEventListener("abort", onCancel);
          },
          { once: true },
        );
      }
      const response = await this.#net.fetch(
        "https://tfgames.site/phpbb3/ucp.php?mode=login",
        {
          redirect: "manual",
          signal: abortController.signal,
        },
      );
      // ignore body
      abortController.abort();
      return response.redirected ? LoginState.LoggedIn : LoginState.NotLoggedIn;
    } catch (_err) {
      return LoginState.Unknown;
    }
  }

  async login(
    username: string,
    password: string,
    cancel?: AbortSignal,
  ): Promise<LoginState> {
    cancel?.throwIfAborted();

    const abortController = new AbortController();
    if (cancel) {
      const onCancel = function onCancel(this: AbortSignal) {
        abortController.abort(this.reason);
      };
      cancel.addEventListener("abort", onCancel, { once: true });
      abortController.signal.addEventListener(
        "abort",
        () => {
          cancel.removeEventListener("abort", onCancel);
        },
        { once: true },
      );
    }

    const loginPageResponse = await this.#net.fetch(
      "https://tfgames.site/phpbb3/ucp.php?mode=login",
      {
        redirect: "manual",
        signal: abortController.signal,
      },
    );
    if (loginPageResponse.redirected) {
      abortController.abort();
      return LoginState.LoggedIn;
    }
    if (!loginPageResponse.ok) {
      abortController.abort();
      throw new Error("Failed to fetch login page");
    }

    const $ = cheerioLoad(await loginPageResponse.text());
    const $form = $("form[id='login']");
    const loginData: Record<string, string> = {
      username,
      password,
      creation_time: $form.find("input[name='creation_time']").val() as string,
      form_token: $form.find("input[name='form_token']").val() as string,
      login: "Login",
      redirect: "/",
      autologin: "1",
    };

    const loginResponse = await this.#net.fetch(
      "https://tfgames.site/phpbb3/ucp.php?mode=login",
      {
        method: "POST",
        body: new URLSearchParams(loginData),
        redirect: "manual",
        signal: abortController.signal,
      },
    );
    abortController.abort();
    return loginResponse.redirected
      ? LoginState.LoggedIn
      : LoginState.NotLoggedIn;
  }
}
