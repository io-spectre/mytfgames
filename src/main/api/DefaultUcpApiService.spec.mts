import * as fs from "node:fs";
import * as path from "node:path";

import { net } from "electron/main";
import { beforeEach, describe, expect, it, Mocked, vi } from "vitest";

import { DefaultUcpApiService } from "./DefaultUcpApiService.mjs";
import { LoginState } from "./UcpApiService.mjs";

describe("DefaultUcpApiService", () => {
  let netMock: Mocked<Pick<typeof net, "fetch">>;

  beforeEach(() => {
    netMock = {
      fetch: vi.fn(),
    };
  });

  it("should be instantiatable", () => {
    const subject = new DefaultUcpApiService(netMock);
    expect(subject).toBeInstanceOf(DefaultUcpApiService);
  });

  describe("instance", () => {
    let subject: DefaultUcpApiService;
    beforeEach(() => {
      subject = new DefaultUcpApiService(netMock);
    });

    it("should check login state", async () => {
      netMock.fetch.mockResolvedValueOnce({
        headers: new Headers(),
        ok: false,
        redirected: true,
        status: 0,
        statusText: "",
        type: "opaqueredirect",
        url: "",
        clone: function (): Response {
          throw new Error("Function not implemented.");
        },
        body: null,
        bodyUsed: false,
      } as Response);

      const result = await subject.checkLoginState();

      // ensure that the fetch call is correct (otherwise the test is useless)
      expect(netMock.fetch.mock.calls).toMatchSnapshot();
      expect(result).toBe(LoginState.LoggedIn);
    });

    it("should correctly fill out the login form", async () => {
      const responseBody = fs.readFileSync(
        path.join(import.meta.dirname, "test/ucp-login.html"),
        "utf8",
      );

      netMock.fetch.mockResolvedValueOnce(new Response(responseBody));
      netMock.fetch.mockResolvedValueOnce({
        headers: new Headers(),
        ok: false,
        redirected: true,
        status: 0,
        statusText: "",
        type: "opaqueredirect",
        url: "",
        clone: function (): Response {
          throw new Error("Function not implemented.");
        },
        body: null,
        bodyUsed: false,
      } as Response);

      const result = await subject.login("username", "password");

      // ensure that the fetch call is correct (otherwise the test is useless)
      // the second fetch snapshot body contains the url search params which
      // tests that the login form has been correctly scraped
      expect(netMock.fetch.mock.calls).toMatchSnapshot();
      expect(result).toBe(LoginState.LoggedIn);
    });
  });
});
