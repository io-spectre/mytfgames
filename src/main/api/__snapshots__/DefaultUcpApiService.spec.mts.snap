// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DefaultUcpApiService > instance > should check login state 1`] = `
[
  [
    "https://tfgames.site/phpbb3/ucp.php?mode=login",
    {
      "redirect": "manual",
      "signal": {
        "aborted": true,
        "reason": [AbortError: This operation was aborted],
      },
    },
  ],
]
`;

exports[`DefaultUcpApiService > instance > should correctly fill out the login form 1`] = `
[
  [
    "https://tfgames.site/phpbb3/ucp.php?mode=login",
    {
      "redirect": "manual",
      "signal": {
        "aborted": true,
        "reason": [AbortError: This operation was aborted],
      },
    },
  ],
  [
    "https://tfgames.site/phpbb3/ucp.php?mode=login",
    {
      "body": [
        [
          "username",
          "username",
        ],
        [
          "password",
          "password",
        ],
        [
          "creation_time",
          "1723120353",
        ],
        [
          "form_token",
          "e96feb7a8d63da0370e4ddb56fd045c1160f5413",
        ],
        [
          "login",
          "Login",
        ],
        [
          "redirect",
          "/",
        ],
        [
          "autologin",
          "1",
        ],
      ],
      "method": "POST",
      "redirect": "manual",
      "signal": {
        "aborted": true,
        "reason": [AbortError: This operation was aborted],
      },
    },
  ],
]
`;
